sdk/identity/azure-identity/inc/azure/identity/detail/token_cache.hpp

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Token cache.
+ *
+ */
+
+#pragma once
+
+#include <azure/core/credentials/credentials.hpp>
+
+#include <chrono>
+#include <functional>
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <string>
+
+namespace Azure { namespace Identity { namespace _detail {
+  /**
+   * @brief Access token cache.
+   *
+   */
+  class TokenCache
+#if !defined(TESTING_BUILD)
+      final
+#endif
+  {
+#if !defined(TESTING_BUILD)
+  private:
+#else
+  protected:
+#endif
+    // A test hook that gets invoked before cache write lock gets acquired.
+    virtual void OnBeforeCacheWriteLock() const {};
+
+    // A test hook that gets invoked before item write lock gets acquired.
+    virtual void OnBeforeItemWriteLock() const {};
+
+    struct CacheValue
+    {
+      Core::Credentials::AccessToken AccessToken;
+      std::shared_timed_mutex ElementMutex;
+    };
+
+    mutable std::map<std::string, std::shared_ptr<CacheValue>> m_cache;
+    mutable std::shared_timed_mutex m_cacheMutex;
+
+  private:
+    TokenCache(TokenCache const&) = delete;
+    TokenCache& operator=(TokenCache const&) = delete;
+
+    static bool IsFresh(
+        std::shared_ptr<CacheValue> const& item,
+        DateTime::duration minimumExpiration,
+        std::chrono::system_clock::time_point now);
+
+    std::shared_ptr<CacheValue> GetOrCreateValue(
```

Comment:
add a comment here that it will return value from cache, or create the entry if ti is not found